### PR TITLE
[DNM] raftstore: let the remaining part inherit old region when split

### PR DESCRIPTION
**Pull request by BusyJay**
ref: #1185 

@siddontang @zhangjinpeng1987 @disksing @hhkbp2 PTAL

/cc @huachaohuang

### CODE REVIEW

File: tests/raftstore/test_split_region.rs [hunk 1]
```diff
@@ -521,7 +521,7 @@ fn test_split_stale_epoch<T: Simulator>(cluster: &mut Cluster<T>) {
     assert!(resp.get_header().has_error());
     assert!(resp.get_header().get_error().has_stale_epoch());
     assert_eq!(resp.get_header().get_error().get_stale_epoch().get_new_regions(),
-               &[left, right]);
+               &[right, left]);
```
Comment (BusyJay) [inline tests/raftstore/test_split_region.rs]: @disksing The order may be change, is it ok?

**Comment by siddontang** [comment]
Seem that you don't change all associated split tests.
I am confused that why https://github.com/pingcap/tikv/blob/398db340aa07ee39b7e852a26374e2ae1d7ecebb/tests/raftstore/test_split_region.rs#L153 can pass?

**Comment by BusyJay** [comment]
When region origin is split into region left and region right, left's start key and right's end key should be same as origin, left's end key should be same as right's start key. The changes is that in the past left's id is the same as origin, but now right's id is the same as origin.

**Comment by siddontang** [comment]
Got it.

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -907,7 +907,8 @@ impl<T: Transport, C: PdClient> Store<T, C> {
                     .is_none() {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: L906 s/new_region_id/right.get_id()

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -1066,11 +1068,11 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             // matter if the next region is not split from the current region. If the region meta
             // received by the TiKV driver is newer than the meta cached in the driver, the meta is
             // updated.
```
Comment (disksing) [inline src/raftstore/store/store.rs]: Please update the comment.

File: tests/raftstore/test_split_region.rs (hunk 1 above)
Comment (disksing) [inline tests/raftstore/test_split_region.rs]: It's okay.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Comment by disksing** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: src/raftstore/store/worker/pd.rs [hunk 4]
```diff
@@ -258,6 +258,10 @@ impl<T: PdClient> Runner<T> {
                       pd_region);
                 PD_VALIDATE_PEER_COUNTER_VEC.with_label_values(&["peer valid"]).inc();
             }
+            Ok(None) => {
+                // split region has not yet report to pd.
+                // TODO: handle merge
```
Comment (hhkbp2) [inline src/raftstore/store/worker/pd.rs]: Two PRs mixed.

File: tests/raftstore/pd.rs [hunk 5]
```diff
@@ -193,7 +192,8 @@ impl Cluster {
         let conf_ver = region.get_region_epoch().get_conf_ver();
         let end_key = enc_end_key(&region);
 
-        let cur_region = self.get_region_by_id(region.get_id()).unwrap();
+        // it can pass handle_heartbeat_version means it must exist.
```
Comment (hhkbp2) [inline tests/raftstore/pd.rs]: Not a good style to assert the binding to another function `handle_heartbeat_version` here, in order to unwrap() directly.

File: src/raftstore/store/store.rs [hunk 6]
```diff
@@ -1062,15 +1063,15 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-            // Attach the next region which might be split from the current region. But it doesn't
-            // matter if the next region is not split from the current region. If the region meta
-            // received by the TiKV driver is newer than the meta cached in the driver, the meta is
-            // updated.
-            if let Some((_, &next_region_id)) = self.region_ranges
-                .range(Excluded(&enc_end_key(peer.region())), Unbounded::<&Key>)
-                .next() {
-                let next_region = self.region_peers[&next_region_id].region();
-                new_regions.push(next_region.to_owned());
+            // Attach the previous region which might be split from the current region. But it
```
Comment (hhkbp2) [inline src/raftstore/store/store.rs]: Why do this attachment? And why `But it doesn't matter`?

File: src/raftstore/store/store.rs [hunk 7]
```diff
@@ -1062,15 +1063,15 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            // doesn't matter if the previous region is not split from the current region. If the
+            // region meta received by the TiKV driver is newer than the meta cached in the
+            // driver, the meta is updated.
```
Comment (hhkbp2) [inline src/raftstore/store/store.rs]: `the meta is updated` means `the cache is updated`? where does this update happen?

File: src/raftstore/store/store.rs (hunk 6 above)
Comment (BusyJay) [inline src/raftstore/store/store.rs]: @disksing

File: tests/raftstore/pd.rs (hunk 5 above)
Comment (BusyJay) [inline tests/raftstore/pd.rs]: Why?

File: src/raftstore/store/worker/pd.rs (hunk 4 above)
Comment (BusyJay) [inline src/raftstore/store/worker/pd.rs]: Yes, I merge the other pr to fix CI.

File: tests/raftstore/pd.rs (hunk 5 above)
Comment (hhkbp2) [inline tests/raftstore/pd.rs]: If the requirement is not true, then this comment is suspicious.

File: src/raftstore/store/store.rs (hunk 6 above)
Comment (disksing) [inline src/raftstore/store/store.rs]: @hhkbp2 It's for this issue: https://github.com/pingcap/tikv/issues/974

**Comment by hhkbp2** [comment]
LGTM

**Comment by BusyJay** [comment]
@huachaohuang any update?

**Comment by ngaut** [comment]
This is not a common solution. what if there are lots of random writing ?

**Comment by huachaohuang** [comment]
This can be merged after https://github.com/pingcap/pd/pull/359 has been merged.

**Comment by BusyJay** [comment]
If it's sequence write, then it behaves better than current implement, otherwise it's not worse than current implement.

**Comment by BusyJay** [comment]
Close it now, will reopen when batch split can't solve this problem.

**Pull request closed without merging**